client/lib/core/utils/media_device_service.dart

```diff
@@ -0,0 +1,188 @@
+import 'dart:async';
+
+import 'package:client/services.dart';
+import 'package:permission_handler/permission_handler.dart';
+import 'package:universal_html/html.dart' as html;
+
+class MediaDeviceService {
+  // Singleton class
+  MediaDeviceService._internal();
+  static final MediaDeviceService _instance = MediaDeviceService._internal();
+  factory MediaDeviceService() => _instance;
+
+  late PermissionStatus micPermissionStatus;
+  late PermissionStatus cameraPermissionStatus;
+
+  // Media stream for local A/V preview
+  html.MediaStream? _previewMediaStream;
+  html.MediaStream? get previewMediaStream => _previewMediaStream;
+
+  List<html.MediaDeviceInfo> audioInputs = [];
+  List<html.MediaDeviceInfo> videoInputs = [];
+
+  String? selectedAudioInputId;
+  String? selectedVideoInputId;
+
+  Future<PermissionStatus> requestPermissions(Permission permission) async {
+    try {
+      // The ".status" call does not work on all platforms - catch the exception.
+      PermissionStatus status = await permission.status;
+      try {
+        status = await permission.status;
+      } catch (e) {
+        return PermissionStatus.denied;
+      }
+      return status.isGranted ? status : await permission.request();
+    } catch (e) {
+      return PermissionStatus.denied;
+    }
+  }
+
+  Future<void> init() async {
+    try {
+      // Start by requesting permissions so that devices can be listed.
+      micPermissionStatus = await requestPermissions(Permission.microphone);
+      cameraPermissionStatus = await requestPermissions(Permission.camera);
+
+      final devices =
+          await html.window.navigator.mediaDevices?.enumerateDevices();
+      if (devices != null) {
```

Comment:
So I think that if there are no available devices, the return value from `enumerateDevices()` will be an empty `Array` rather than `null`. [MDN says the resolved return value is always an array of `MediaDeviceInfo` objects](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices#return_value), which implies to me it would be an empty array if there were no devices. And then if we look at [the specification for `enumerateDevices`](https://w3c.github.io/mediacapture-main/#dom-mediadevices-enumeratedevices), step 12 is always to return a _resultList_, which in our context is an `Array`.

So I guess my suggestion is:

```suggestion
      if (devices?.length > 0) {
```

with the optional chaining operator to catch cases where this assumption is wrong or misimplemented.

I would test this myself but I am not sure how to emulate a device with no camera or microphone!